open_spiel/python/algorithms/projected_replicator_dynamics.py

```diff
@@ -73,6 +73,48 @@ def _project_distribution(updated_strategy, gamma):
   return updated_strategy
 
 
+
+def _project_distribution(updated_strategy, gamma):
+  """Projects the distribution in updated_x to have minimal probabilities.
+
+  Minimal probabilities are set as gamma, and the probabilities are then
+  renormalized to sum to 1.
+
+  Args:
+    updated_strategy: New distribution value after being updated by update rule.
+    gamma: minimal probability value when divided by number of actions.
+
+  Returns:
+    Projected distribution.
+  """
+  # Epsilon approximation of L2-norm projection onto the Delta_gamma space.
+  updated_strategy[updated_strategy < gamma] = gamma
+  updated_strategy = updated_strategy / np.sum(updated_strategy)
+  return updated_strategy
+
+def _project_on_clipped_simplex(v, gamma=0):
+    """
+    Project v on the closest point in L2-norm on gamma-simplex 
+    Based on: https://eng.ucmerced.edu/people/wwang5/papers/SimplexProj.pdf
+    
+    Args:
+      v: New distribution value after being updated by update rule.
+      gamma: minimal probability value when divided by number of actions.
+    
+    Returns:
+      Projected distribution
+    """
+    
+    N = len(v)
+    l = np.sort(v)
+    lam = (np.sum(l)-1)/N
+    for i in range(N-1, 0, -1):
+        pho = (1/(N-i)) * (np.sum(l[i:])-1+gamma * i)
+        if(l[i-1] - pho < gamma):
+            lam = pho
+            break
+    return np.maximum(v-lam, gamma)
```

Comment:
You could try to vectorise your solution to eliminate the for-loop. I think for large strategy spaces a `numpy` version will be faster. Quick sketch below based on the paper reference you provided:

```
def simplex_projection(y, gamma=0.0):
  n = len(y)
  idx = np.arange(1, n + 1)
  u = np.sort(y)[::-1]
  u_tmp = (1 - np.cumsum(u) - (n - idx) * gamma) / idx
  rho = np.searchsorted(u + u_tmp <= gamma, True)
  return np.maximum(y + u_tmp[rho - 1], gamma)
```